translator/internaldata/oc_to_metrics.go

```diff
@@ -0,0 +1,349 @@
+		startTimestamp := internal.TimestampToUnixNano(timeseries.GetStartTimestamp())
+
+		for _, point := range timeseries.GetPoints() {
+			pointTimestamp := internal.TimestampToUnixNano(point.GetTimestamp())
+			labelKeyValues := getLabelKeyValues(ocLabelsKeys, timeseries.GetLabelValues())
+			switch point.Value.(type) {
+			case *ocmetrics.Point_Int64Value:
+				dataPoints := metric.Int64DataPoints()
+				if dataPoints.Len() == 0 {
+					dataPoints.Resize(ocPointsCount)
+				}
+				dataPoint := dataPoints.At(int64DataPointsNum)
+				dataPoint.SetStartTime(startTimestamp)
+				dataPoint.SetTimestamp(pointTimestamp)
+				setInt64DataPointValue(dataPoint, point)
+				setDataPointLabels(dataPoint, labelKeyValues)
+				int64DataPointsNum++
+			case *ocmetrics.Point_DoubleValue:
+				dataPoints := metric.DoubleDataPoints()
+				if dataPoints.Len() == 0 {
+					dataPoints.Resize(ocPointsCount)
+				}
+				dataPoint := dataPoints.At(doubleDataPointsNum)
+				dataPoint.SetStartTime(startTimestamp)
+				dataPoint.SetTimestamp(pointTimestamp)
+				setDoubleDataPointValue(dataPoint, point)
+				setDataPointLabels(dataPoint, labelKeyValues)
+				doubleDataPointsNum++
+			case *ocmetrics.Point_DistributionValue:
+				dataPoints := metric.HistogramDataPoints()
+				if dataPoints.Len() == 0 {
+					dataPoints.Resize(ocPointsCount)
+				}
+				dataPoint := dataPoints.At(histogramDataPointsNum)
+				dataPoint.SetStartTime(startTimestamp)
+				dataPoint.SetTimestamp(pointTimestamp)
+				setHistogramDataPointValue(dataPoint, point)
+				setDataPointLabels(dataPoint, labelKeyValues)
+				histogramDataPointsNum++
+			case *ocmetrics.Point_SummaryValue:
+				dataPoints := metric.SummaryDataPoints()
+				if dataPoints.Len() == 0 {
+					dataPoints.Resize(ocPointsCount)
+				}
+				dataPoint := dataPoints.At(summaryDataPointsNum)
+				dataPoint.SetStartTime(startTimestamp)
+				dataPoint.SetTimestamp(pointTimestamp)
+				setSummaryDataPointValue(dataPoint, point)
+				setDataPointLabels(dataPoint, labelKeyValues)
+				summaryDataPointsNum++
+			}
+		}
+	}
+}
+
+func getLabelKeyValues(ocLabelsKeys []*ocmetrics.LabelKey, ocLabelValues []*ocmetrics.LabelValue) []labelKeyValue {
+	if len(ocLabelsKeys) == 0 || len(ocLabelValues) == 0 {
+		return nil
+	}
+
+	lablesCount := len(ocLabelsKeys)
+
+	// Handle invalid length of OC label values list
+	if len(ocLabelValues) < lablesCount {
+		lablesCount = len(ocLabelValues)
+	}
+
+	labelKeyValues := make([]labelKeyValue, 0, lablesCount)
+	for i := 0; i < lablesCount; i++ {
+		if !ocLabelValues[i].GetHasValue() {
+			continue
+		}
+		labelKeyValues = append(labelKeyValues, labelKeyValue{
+			Key:   ocLabelsKeys[i].Key,
+			Value: ocLabelValues[i].Value,
+		})
+	}
+
+	return labelKeyValues
+}
+
+func setDataPointLabels(dataPoint labelsDataPoint, labels []labelKeyValue) {
+	if len(labels) == 0 {
+		return
+	}
+
+	labelsMap := dataPoint.LabelsMap()
+	for _, label := range labels {
+		labelsMap.Insert(label.Key, label.Value)
+	}
+}
```

Comment:
You can pass directly dataPoint.LabelsMap().

Reply:
Thanks for the tip! Updated code